[2단계 - 행운의 로또 미션] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
## 🍊 `Step2` 니야의 로또 웹 미션 구현하기 🍊
안녕하세요, 동키콩! 니야입니다. 이번 삼일절 연휴는 잘 보내고 계신가요 😃 
이번 로또 미션 2단계는 웹을 구현하는 미션이었는데, 어떻게 설계해야 효율적으로 구현할 수 있을지 고민이 많이 되더라구요.
그래서 지난 기수 크루들이 구현한 로또 웹 미션을 열심히 찾아보면서 학습하여 제가 도입하고자 하는 설계 방식으로 구현해봤습니다! 
아래는 코드 이해하기 쉽도록 웹 구현 내용에 대해 간단하게 시각화한 플로우 차트입니다.
<img width="866" alt="image" src="https://github.com/user-attachments/assets/88aad192-d4e4-4aa8-b32d-fb3b40455b82" />
`console`과 함께 작업한 미션이다보니 폴더 구조를 어떻게 잡으면 좋을지 더 고민했던 것 같습니다. `console` 기반으로 구현할 때는 `MVC`를 사용했는데요. 그렇기에 웹 개발에서도 동일한 `domain` 로직을 사용하려다보니 `MVC` 구조를 최대한 활용해보고 싶었습니다. 그래서 controller와 `view`를 사용했습니다. `controller`에서는 전반적인 로직 관리를 중점적으로 했고, 여기서 `domain`과 `view`를 통신해주는 역할을 했습니다. `view` 에서는 전체 구조를 담은 `View` 클래스를 하나 만들어두고, 세부 내용은 코드 볼륨도 커지고 폴더 `depth`가 깊어질 것 같아서 `components` 폴더를 새로 생성하여 독립적으로 관리해주었습니다. 이번에 `customElements`를 사용하였는데요. 그 이유는 최대한 캡슐화 하여 각각 모듈별로 응집도를 높이고 싶었기 때문입니다. 그래서 크게 총 6가지의 커스텀 컴포넌트로 분리하여 묶어서 관리해봤습니다. 여기서 `HTMLElement`를 상속받은 `BaseWebComponent` 라는 추상 클래스를 하나 생성했는데요. 각각의 컴포넌트들이 `BaseWebComponent`를 상속받아서 공통적으로 사용하는 메서드를 사용함으로써, 코드 재사용성이나 일관성을 높일 수 있도록 해주고 싶었습니다.

## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://sooyeoniya.github.io/javascript-lotto/

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 

- 이번에는 어떻게 보면 `view`라고 볼 수 있는 `component`의 내부에서 `입력에 대한 유효성 검증 함수(domain)`를 직접 호출하는 방식으로 진행했습니다. `console` 기반 로또 미션 진행 당시에는 `MVC` 패턴에 대한 생각이 확고하여 최대한 각각을 독립적으로 분리하려고 노력했었습니다. 그리고 우테코에서도 `domain`과 `UI` 를 최대한 분리하라고 권장했구요. 근데 이번에 `domain`이라고 생각했던 유효성 검증 로직에 대한 생각이 많이 바뀌었습니다. 지난 수업시간에 로또 웹 개발에 대한 시퀀스 다이어그램을 그려보는 과정에서 공원한테 배운 게 있었습니다. 입력 유효성 검증 로직은 `UI`에서도 진행할 수 있더라구요. 특히 이제 웹으로 넘어오면 `UI`즉 `view`가 담당하는 부분이 더 많아질 것이라고 말씀해주셨습니다. 듣고 보니 공원의 말씀이 확실히 체감되었습니다. 실제로 입력은 이제 `UI`에서 받아서 에러가 발생하면 바로 화면에 표시해주어야하는데, 유효성 검증을 그 자리에서 바로 해주면 되지, 굳이 불필요한 경로를 만들 필요는 없다고 생각했습니다. 예를 들어, 아래 `WinningLotto` 객체에 있는 메서드는, 당첨 번호와 보너스 번호를 받아 에러 메시지를 `UI`에 직접 출력해주는 메서드입니다. 근데 이 부분을 `domain`인 `validate`메서드들과 `view`인 `hideElement`, `renderElement` 이런 부분들과 독립적으로 작성하기에는 좀 어려운 부분이 있어서 그냥 `WinningLotto` 객체 내부에서 해주는 것이 맞다고 판단했습니다. 왜냐하면 `controller`까지 갔다가 유효성 검증된 값을 받아오기에는 너무 불필요한 경로 이동이 존재하여 더 코드가 복잡해질 것 같다고 생각했기 때문입니다. 그래서 유효성 검증을 `UI` 출력 부분과 같은 곳에서 한 번에 작업해주는 것이 더 효율적이라고 생각하여 이리 구현하게 되었습니다.
```js
  #handleValidation(winningNumbersInput, bonusNumberInput, errorElement) {
    try {
      const winningNumbers = validateWinningNumbers(winningNumbersInput);
      const bonusNumber = validateBonusNumber(bonusNumberInput, winningNumbers);
      hideElement(errorElement);
      this.emit(EVENT_TYPES.result, { winningNumbers, bonusNumber });
    } catch (error) {
      errorElement.textContent = error.message;
      renderElement(errorElement);
    }
  }
```

### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

1️⃣ 전반적인 컴포넌트 명, 이벤트 타입 명 등에 대해서 이상한 부분이 있다면 피드백 부탁드립니다!

2️⃣ UX 측면에서 좀 불편한 부분이 있다면 피드백 부탁드립니다.

3️⃣ `View`객체의 위치가 애매합니다. `View`에는 전체 구조를 초기화해주는 곳인데, 사실상 `render` 메서드만 쓰기 때문에, `BaseWebComponent`를 굳이 상속받을 필요가 없어서 그냥 따로 독립적으로 두긴 했습니다. 하지만, `views/web/View.js` 얘만 혼자 동떨어져있는 느낌도 듭니다. `components` 폴더 내부에 함께 넣는게 맞는 걸까요? 그렇지만, `MVC` 패턴을 위해서 만든 폴더 구조가 깨지는 느낌이라 지금처럼 그냥 두고싶기도 합니다.

4️⃣ 전반적인 설계 방식과 전체 흐름에 대해서 피드백 받고 싶습니다. 예를 들어, `BaseWebComponent` 에서 상속받을 메서드가 잘못 정의되어있다든지, 메모리 누수 방지를 위해 중간에 이벤트 리스너 제거가 필요한 부분이 있다든지, 렌더링 위치가 성능 문제에 영향을 준다든지 등등 그런 부분에서 만약 피드백이 있다면 해주시면 감사하겠습니다. 그리고 특히 `사용자 정의 요소(customElements)`는 이번에 처음 접근해본 방식이어서 제대로 잘 진행된건지, 제가 혹시 놓치는 게 있는지 궁금합니다. 

5️⃣ `customElements`를 상수화하는 과정에서 고민이 생겼습니다. 아래와 같이 총 6개의 컴포넌트 태그명을 상수화하였는데요. 
```js
export const CUSTOM_ELEMENTS = Object.freeze({
  lottoHeader: "lotto-header",
  lottoPurchase: "lotto-purchase",
  issuedLotto: "issued-lotto",
  winningLotto: "winning-lotto",
  lottoResult: "lotto-result",
  lottoFooter: "lotto-footer",
});
```
이걸 상수화 한 이유가 `LottoController`에서 이 각 커스텀 태그들을 접근해야하다보니 하드코딩 되어있었습니다. 따라서 이렇게 하드코딩 된 부분을 없애고자 상수를 사용하자는 생각이 있었습니다. 사실 상수의 목적이 유지보수성을 높이기 위해서인데, 진짜 `유지 보수`가 참된 목적이라면 어디까지 상수화를 해야하는지 갑자기 미궁에 빠지게 되었습니다. 하나의 커스텀 컴포넌트만 살펴봐도 알 수 있습니다. 아래는 `LottoPurchase` 객체의 요소들입니다. 이것의 사용자 정의 요소 태그명은 위에서 찾을 수 있다시피 `lotto-purchase` 입니다. 그런데 내부에 `class` 선택자 이름들의 `prefix`를 모두 `lotto-purchase`로 통일해놨는데요. 이것까지도 다 상수화해줘야하는 부분일까요? 그래야만 유지보수하기 수월해질 수 있을텐데, 너무 가독성이 떨어질 것 같아서 수정하지 않았습니다.

```js
<section class="lotto-purchase">
  <h2 class="lotto-purchase__title">🎱 내 번호 당첨 확인 🎱</h2>
  <label for="purchase-amount" class="lotto-purchase__description">구입할 금액을 입력해주세요.</label>
  <form class="lotto-purchase__form">
    <input id="purchase-amount" class="lotto-purchase__input" placeholder="금액" />
    <button class="lotto-purchase__button">구입</button>
  </form>
  <p class="lotto-purchase__error ${STYLE_SELECTORS.hidden}"></p>
</section>
```
아래 부분도 마찬가지입니다. 아래는 `View` 객체의 요소들인데요. 이는 전체 커스텀 컴포넌트 구조를 정의해주는 부분입니다. 그런데 이것들까지 전부 상수화하기에는 위에서 말했던 것과 마찬가지로 가독성이 매우 떨어질 것 같아서 그냥 이대로 두었습니다. 
```js
render() {
    this.app.innerHTML = `
    <lotto-header></lotto-header>
    <div class="container">
      <main>
        <lotto-purchase></lotto-purchase>
        <issued-lotto></issued-lotto>
        <winning-lotto></winning-lotto>
        <lotto-result></lotto-result>
      </main>
    </div>
    <lotto-footer></lotto-footer>
    `;
  }
```
상수화의 목적은 재사용성과 유지보수성인데, 그냥 `LottoController`에서 단순히 태그명들이 하드코딩되었다고 상수화하는 것은 옳지 않은 방식이었던 걸까요? 유지보수까지 고려하면 너무 많은 것을 상수화해야하는 것 같아서 이건 아닌 것 같다는 생각이 들었습니다. 보통 실무에서는 어떻게 상수를 활용하는지 그 기준이 궁금합니다. 제 입장이라면 동키콩은 어떤 선택을 하실지 궁금합니다! 

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

sooyeoniya committed:
refactor: InputService -> InputController로 변경

- Service가 아닌 Controller의 역할이 맞다고 판단

sooyeoniya committed:
chore: console 폴더로 위치 변경

- Controller와 View 부분 console 폴더 하위로 이동
- web 폴더와 구분하기 위함

sooyeoniya committed:
feat(UI): 로또 구매 및 발행, 당첨 번호에 대한 웹 UI 및 CSS 스타일 추가

- 로또 구입 금앱 입력
- 로또 발행된 부분
- 로또 당첨 번호 및 보너스 번호 입력
- 헤더 및 푸터

sooyeoniya committed:
refactor(CSS): CSS 폴더 분리

- 각 컴포넌트 별로 분리
- 로또 구입하는 부분, 발행한 로또 출력하는 부분, 당첨 번호 및 보너스 번호 입력 받는 부분

sooyeoniya committed:
feat(UI): 로또 통계 결과 모달창 구현

- 모달창 Html, CSS 구성
- favicon 추가

sooyeoniya committed:
style(UI): 버튼 active, hover 시 색상 변경

sooyeoniya committed:
chore: css 폴더 분리

- footer, header, reset 폴더 생성 및 각 역할에 맞게 분리

sooyeoniya committed:
feat(BaseWebComponent): 커스텀 웹 컴포넌트를 위한 추상 클래스 구현

sooyeoniya committed:
fix: shaow DOM 제거

- css 스타일 적용 안됨 이슈

sooyeoniya committed:
feat: 컴포넌트 구조 설계

- 컴포넌트를 분리하여 각 js와 css 파일을 한 번에 묶어서 관리
- LottoController 객체 인스턴스 실행 시 view 생성

sooyeoniya committed:
feat: 로또 구입 및 발행 구현

- 입력폼에 구입 금액 입력 시 그만큼의 로또가 발행됨
- 구입 금액의 입력값이 잘못된 경우 에러 출력 후 다시 입력받을 수 있도록 함

sooyeoniya committed:
feat: 당첨 번호 및 보너스 번호 웹 구현

- 입력폼에 당첨번호와 보너스 번호 입력 후 유효성 검증 연동
- 유효하지 않은 경우 화면에 에러 출력

sooyeoniya committed:
feat: 당첨 내역 결과 출력 구현

- 통계 및 수익률 값 UI와 연동

sooyeoniya committed:
refactor(WinningLotto): 동일한 로직을 부모 클래스의 메서드로 호출 변경

- render와 setEvent를 묶어준 connectedCallback으로 대체

sooyeoniya committed:
feat: dialog를 사용해 모달창 구현

- CSS 파일에서 close_button 부분의 언더바를 하이픈(-)으로 변경
- 모달 닫기 버튼 구현 -> 전체 렌더링 되지 않도록 method='dialog' 추가
- dialog는 초기상태가 닫혀있어야하는데, css 스타일 적용으로 닫혀있지 않았음, 이에 내부에 form 태그 추가하여 수정

sooyeoniya committed:
feat: 게임 재시작 로직 연동

- this.setEvent()를 다시 호출안해줘서 버튼 클릭이 제대로 되지 않는 오류 수정
- LottoController에서 게임 재시작 할 때 view를 다시 렌더링 해줌

sooyeoniya committed:
feat(LottoPurchase): 구입 금액 label 태그로 변경

sooyeoniya committed:
feat(WinningLotto): 당첨번호 및 보너스 번호 2개 입력 시 다음 input으로 이동

sooyeoniya committed:
feat(LottoResult): 모달창 닫힐 때 이벤트 리스너 제거

- 모달창의 X 버튼 클릭 시 모달창이 닫힐 때 이벤트 리스너를 제거하는 것이 과연 괜찮은 방식인걸까?
- X 버튼에서는 이벤트 리스너를 제거하지 않고, 게임 재실행할때만 제거하는 것이 좋은 방식일까?

sooyeoniya committed:
refactor(LottoResult): 이벤트 등록 및 제거 공통 로직 분리

- this.on.bind(this), this.off.bind(this)로 사용한 이유: 현재 클래스 인스턴스의 컨텍스트로 바인딩 하기 위함
- 만약 JavaScript에서 메서드를 객체에서 분리해 호출하는 경우 this가 undefined가 되거나 전역 객체를 가리킬 수 있으므로 bind 메서드를 사용해 명확하게 바인딩해주도록 함

sooyeoniya committed:
refactor(domUtils): 에러 메시지 요소 추가 및 제거 로직 수정

- 목적: JS 와 CSS 관심사 분리
- style.display 스타일을 JS 에서 직접 수정하고 있었음.
- CSS에게 직접 책임 이양하기 위해 hidden 클래스 생성하여 클래스 추가 및 제거를 통해 에러 메시지를 제어함
- domUtils에 hideElement, renderElement를 만들어 재사용하도록 함

sooyeoniya committed:
refactor(WinningLotto): 이벤트 리스너 등록 메서드 분리

sooyeoniya committed:
refactor: 에러 메시지 try-catch 부분 메서드 분리

sooyeoniya committed:
refactor(domUtils): querySelector, querySelectorAll 랩핑하는 유틸함수 정의

- 코드 가독성 및 간결성 향상을 위함

sooyeoniya committed:
refactor(WinningLotto): 당첨 번호 및 보너스 번호 반환 메서드 분리

sooyeoniya committed:
refactor(domUtils): BaseWebComponent의 on, off 메서드 domUtils로 이동

- BaseWebComponent에서 상속받을 필요 없는 메서드라고 판단하여 유틸함수로 정의함

sooyeoniya committed:
refactor(LottoController): 웹 로또 컨트롤러 함수 분리

sooyeoniya committed:
refactor(constants): 상수 파일 분리

- 상수가 많아서 파일 분리하는 것이 관리에 수월하다고 판단함

sooyeoniya committed:
refactor(constants): 이벤트 타입 및 hidden 클래스 선택자 상수화

- 이벤트 타입 상수 정의 및 hidden 선택자 상수 정의
- 추가 작업: 상수 파일 이름 suffix 'Constants' 삭제

sooyeoniya committed:
chore: 콘솔 컨트롤러 클래스명 변경

- Controller -> LottoController

sooyeoniya committed:
refactor(constants): custom element 태그명 상수화

- LottoController에서 사용되는 custom element 하드코딩 된 부분이 있어서 상수화함
- 이에 따라 각각의 custom 컴포넌트에서도 상수 적용

궁금한 점
- View에 있는 태그들도 다 상수화하는 게 맞을까? -> 복잡도가 올라갈 것 같음
- 하나의 custom 컴포넌트 내부에 있는 여러개의 하위 태그들에 대하여 클래스명의 prefix가 custom element 상수명과 동일한데 이것도 상수화해야하는걸까? -> 이것 또한 조잡해질 것 같음

- 하지만 위의 것들을 진행했을 경우 유지보수 측면에서 괜찮을 것 같다는 생각.

결론
- 도대체 어디까지 상수화를 하는 것이 적절한 것인가.

sooyeoniya committed:
build: git pages 배포 경로 수정

sooyeoniya committed:
docs: README.md 기능 요구사항 작성 및 PR 문서 내용 docs 폴더에 저장

JUDONGHYEOK commented on src/step2-index.js:
```diff
@@ -2,3 +2,8 @@
  * 노드 환경에서 사용하는 readline 등을 불러올 경우 정상적으로 빌드할 수 없습니다.
  */
+
+import "./styles/global.css";
```
js파일에서 css를 Import하셨네요. html파일에서 link로 가져오는 것과 어떤 차이가 있을까요?

sooyeoniya replied:
우선 각 컴포넌트마다 CSS를 import 해주었는데 그 이유는 스타일링 응집성 높이기 위해 각 컴포넌트 단위로 CSS를 관리해주었습니다! 근데 최상위에서 global.css 를 직접 import 하는 것과 HTML에서 link 태그로 가져오는 것에 대해서는 깊게 고민하지 않고 사용했던 것 같네요.

학습해보니, 크게 `로드 순서`, `모듈화`, `번들링` 관련 부분에서 차이가 있었습니다!

1. `로드 순서`: import 해주는 방식은 JS 파일에 의해 CSS 파일이 로드되기 때문에 FOUC 같은 문제가 발생할 수 있고, link 태그 방식은 HTML 파일 로드 시 CSS 파일이 함께 로드되기 때문에 스타일이 즉시 적용되어 FOUC 문제를 방지할 수 있습니다
2. `모듈화`: import 방식은 CSS 파일을 각 컴포넌트나 모듈 단위로 스타일 관리가 가능했고, link 태그 방식은 모듈화가 어렵고 해당 CSS가 모든 페이지에서 로드되기 때문에 전역 스타일을 적용할때 유리합니다!
3. `번들링`: 빌드도구를 통한 번들링 과정에서도 차이가 있었는데요. import 방식은 CSS 가 JS 번들에 포함되어 있기 때문에 번들 크기가 커질 수도 있지만, 한 편으로는 트리셰이킹 같은 사용되지 않는 CSS 코드를 삭제할 수 있기 때문에 최적화가 가능합니다. link 태그 방식은 CSS 파일을 별도로 관리해야 해서 이런 부분에 있어서는 빌드 도구와의 통합이 어렵다는 단점이 있네요.

결론적으로 Vite 빌드툴을 사용하고 있기 때문에 모듈화가 가능해서 각 컴포넌트 단위로 CSS 파일을 내부에 직접 import 해서 관리해주는 것은 좋은 방식이라고 생각했습니다. 다만 지금 global.css 내부에 @import 방식으로 reset.css를 불러오고 있는데 브라우저 호환성 측면이나 페이지 로딩 시간 관련해서 문제가 있을 수 있다고 해서 global과 reset파일은 link 태그 방식으로 관리해주는 것이 더 괜찮은 방식인 것 같기도 하네요! 생각할거리 주셔서 정말 감사합니다 ! 덕분에 당연시 여기던 것도 한 번 더 찾아보고 깊이 학습해 볼 수 있는 시간이 되었던 것 같아용 😁

https://github.com/woowacourse/javascript-lotto/pull/406/commits/ffdb580572e8901277191735a5f11b547f3846c3

JUDONGHYEOK replied:
오 최고입니다! 니야 차이점에 대해서 명확하게 정리해주셨네요! 판단해주신 결정에도 동의합니다!

JUDONGHYEOK commented on src/controllers/web/LottoController.js:
```diff
@@ -0,0 +1,79 @@
+class LottoController {
+  constructor() {
+    this.lottos = [];
+    this.purchaseAmount = 0;
```
controller에서 해당 상태를 관리하는 것이 맞을까 생각이 되요. controller의 목적이 model과 view의 관계를 중재하는 역할이라면 Controller에서 도메인성 상태를 갖고있는 것이 객체의 역할을 벗어난게 아닌가 생각이 됩니다!

또 view를 여기서 생성하기 보다는 외부에서 주입받는 것이 결합도를 낮추는데 도움이 될 것 같아요

sooyeoniya replied:
동키콩 말씀대로 MVC 패턴을 사용한다고는 했지만 명확히 역할이 구분된 건 아닌 것 같았네요! 컨트롤러에서 lottos나 purchaseAmount 같은 상태를 관리하는 것이 맞지 않는다는 생각이 들었어요. 따라서 웹을 위한 도메인 객체를 새로 만들어서 공통으로 사용되는 도메인 로직을 composition 하여 사용하도록 해주었습니다. 이는 웹 컨트롤러와 통신하기 위한 상태 관리하는 도메인 객체라고 보면 될 것 같아요! 그리고 컨트롤러 내부에서 View를 직접 호출하는 것보다 외부에서 주입하는 것이 훨씬 결합도를 낮추는 일이라는 것을 알게 되었습니다. 추가로 View와 관련된 메서드는 모두 View에서 관리되도록 하여 각자의 역할을 잘 수행하도록 해주었습니다! 수정된 커밋 내역은 다음과 같습니다. https://github.com/woowacourse/javascript-lotto/pull/406/commits/7ef9aeec3098e3f6a81f862dbfb9d2b1114a8bfb

JUDONGHYEOK replied:
오 좋아요 controller의 의존성이 확실히 줄어든 것 같습니다! MVC패턴이 꼭 정답은 아니지만 해당 패턴 도입이 필요하다고 느끼고 도입을 할때는 항상 controller가 비대해지는 것을 경계해야하는 것 같아요. 수정버전이 확실히 깔끔하게 구조화된 것 같습니다!

JUDONGHYEOK commented on src/utils/domUtils.js:
```diff
@@ -0,0 +1,25 @@
+
+export const $ = (selector, target = document) => {
+  return target.querySelector(selector);
+};
```
해당 select코드는 표준 DOMAPI에 querySelector가 사용하기 전에 jQuery에서 사용되던 방식인 것 처럼 보이는 데요. 사실 querySelector가 표준 DOMAPI에 포함되면서 유틸화하여 사용하는 것에 대해 큰 이점이 있을까 싶어요. 참조를 하나 더 만드는 느낌이 드는 것 같은데 어떻게 생각하세요?

sooyeoniya replied:
저도 다른 여러 코드를 참고하여 유틸화 해준 부분인데, 사실 그 필요성에 대해서 깊게 고민해보지 않았던 것 같네요. 실질적으로 querySelector() 그 자체만으로도 충분히 직관적이고 간결한 것 같아서 단순히 참조 하나만 더 만드는 느낌이 들 수도 있겠네요! $나 $$로 추상화했을 때 단순히 코드만 짧아질 뿐이지, 어떤 것을 의미하는지 다시 까봐야 알기 때문에 오히려 가독성이 더 좋지 않을 수도 있을 것 같다는 생각이 들었습니다. quertSelector라는 표준 API를 한꺼풀만 덧씌우는 것 자체가 큰 이점이 되는 것 같지는 않아서 기존 상태로 다시 돌려놓았습니다! https://github.com/woowacourse/javascript-lotto/pull/406/commits/3e6a90857af10d3d1af3a79f53c6ab706baef90d

JUDONGHYEOK replied:
코드 스타일의 차이이고 개인의 취향차이일 것 같은데요. 개인적으로 jQuery가 익숙하신 분들은 편하게 사용하실 것 같지만 저는 그렇지 않다보니 그렇게 느낀것 같아요. 니야도 반영해주신걸보니 똑같이 느꼈을거라 생각해요ㅎㅎ 감사합니다 니야!

JUDONGHYEOK commented on src/components/winning-lotto/WinningLotto.js:
```diff
@@ -0,0 +1,130 @@
+  getTemplate() {
+    if (!this.isInitialized) {
+      return "";
+    }
```
플래그변수에 따라 내부요소 렌더링 요소를 결정하기보다 view에서 show hide로직을 관리하고 controller에서 명령하는 형식으로 렌더링여부를 결정하는 것은 어떻게 생각하세요? 컴포넌트의 렌더링 여부를 결정하는 요소가 컨트롤러와 컴포넌트 내부에 숨어있다보니 로직이 한눈에 파악하기 쉽지 않은 것 같아서요! 필요할때 template을 그대로 렌더링 하는것이 더 가독성이 좋지 않을까요?

sooyeoniya replied:
맞는 말씀입니다. 동키콩 말씀대로 해야 view와 controller의 역할이 더 명확해질 것 같네요! 템플릿 자체는 그냥 내부에서 관리하되, 플래그 변수는 외부에서 관리하는게 더 좋을 것 같기도 하네요. 그리고 플래그 변수보다 hidden 클래스를 붙여서 사용하는 것이 더 깔끔한 코드 같습니다! 그래서 기존에 있던 domUtils에 있는 renderElement와 hideElement 메서드를 사용해서 view에서 렌더링 여부를 직접 관리하도록 수정했습니다! 그리고 controller에서 view의 render를 직접 호출하는 것이 아니라 view의 로또 재실행 메서드를 호출해서 그 내부에서 렌더링이 알아서 되도록 선언형 방식으로 바꿨습니다!
https://github.com/woowacourse/javascript-lotto/pull/406/commits/a22f4547ccf4cf6db8671b59f600355e5a99f06f

JUDONGHYEOK replied:
오 좋네요 역할의 분리가 명확해지고 코드의 흐름 파악이 수월해진 것 같아요!👍 관리되어야하는 상태값에 대해서 어느 곳에 위치해야하는 것은 상황에 따라 다르지만 현재로직에서는 플래그변수보다 지금구조가 깔끔한 것 같다고 생각해요!! 감사합니다 니야!

JUDONGHYEOK commented on index.html:
```diff
@@ -1,16 +1,14 @@
     <title>🎱 행운의 로또</title>
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-    <title>Document</title>
+    <link rel="icon" type="image/svg+xml" href="favicon.png" />
```
오 중복되는 타이틀을 없에고 favicon까지 넣어주셨군요!👍

sooyeoniya replied:
그런데 왜 배포한 데모 페이지에서는 favicon이 제대로 적용되지 않는걸까용.. 이 부분도 한 번 수정해보겠습니다

JUDONGHYEOK commented on index.html:
```diff
@@ -1,16 +1,14 @@
-<!DOCTYPE html>
+<!doctype html>
```
이건 왜 소문자로 바꿔주셨는지 이유가 있으실까요??

sooyeoniya replied:
따로 이유는 없고, prettier 적용되는 과정에서 소문자로 변환되었습니다! 

HTML5 표준에서 DOCKTYPE 선언할 때 [대소문자 구분을 안한다(case insensitive)](https://www.scaler.com/topics/html-is-case-sensitive-language/)고 해서 아마 prettier가 코드 스타일 일관성 유지하려고? 소문자로 변환시킨 것 같습니다. 

prettier 제 로컬 버전이 `"version": "3.4.2"` 인데, [Prettier 3.4: A lot of bug fixes](https://prettier.io/blog/2024/11/26/3.4.0#html) 여기서 보면 확장자가 .html인 경우 소문자로 변환된다고 명시되어 있네요.
```md
Starting with Prettier 3.4, we'll only lowercase HTML5 doctype (<!doctype html>) when the file extension is .html or .htm, otherwise they will be untouched. 
```

JUDONGHYEOK commented on src/components/lotto-result/LottoResult.js:
```diff
@@ -0,0 +1,109 @@
+      <dialog>
+        <form class="lotto-result" method="dialog">
+          <button class="lotto-result__close-button">
+            <img src="close-button.svg" alt="close-button" />
```
얘가 로컬에서 실행할때 기준으로 정상적으로 로드되지 않고 있는 것 같아요. 어떻게 개선할 수 있을까요?

sooyeoniya replied:
정상적으로 로드가 잘 안되나요?! 제 로컬과 데모 페이지에서는 당첨 통계 내역이 잘 로드되는 걸 확인했는데, 어떤 부분에서 문제가 있는건지 여쭤봐도 되겠습니까..?

JUDONGHYEOK replied:
제가 뭘 잘못봤나보네요. 죄송합니다 현재는 잘 로드되는 것 확인했습니다!

JUDONGHYEOK commented on src/components/base/BaseWebComponent.js:
```diff
@@ -0,0 +1,36 @@
+    this.setEvent();
+  }
+
+  disconnectedCallback() {
+    this.removeEvent();
+  }
```
현재 컴포넌트가 DOM에서 커스텀엘리먼트가 제거되는 경우는 restart할때만 인것이죠??

sooyeoniya replied:
네 맞습니다. 커스텀엘리먼트가 제거되는 경우는 로또 재실행(restart)되는 경우만 해당됩니다! 컴포넌트가 DOM에 언제 포함되고 포함 안되는지에 대해서는 어떤식으로 로직을 추가해야하는지 아직 감이 잘 안와서 좀 더 생각해보겠습니다!

JUDONGHYEOK commented on src/components/lotto-result/lotto-result.css:
```diff
@@ -0,0 +1,66 @@
+.lotto-result {
+  z-index: 100;
```
지금은 괜찮지만 z-index 같은 아이들도 변수로 한군데서 관리되면 좋을 것 같아요. 애플리케이션이 늘어날수록 관리가 까다로워져서요

sooyeoniya replied:
확실히 애플리케이션이 커질수록 z-index 값이 여러곳에서 중구난방으로 정의되니까 우선순위 충돌이 발생할 수 있겠네요. 한 곳에서 관리해두면 유지보수하기 쉬울 것 같아요! 좋은 말씀 감사합니다!!

JUDONGHYEOK commented on src/controllers/web/LottoController.js:
```diff
@@ -0,0 +1,79 @@
+    lottoResult.showResult(winningStatistics.statistics, profitRatio);
+  }
+
+  #handleRestart() {
+    this.view.render();
+  }
```
재시작하는 비용이 큰게 아닌가 생각되네요? header나 footer같은 경우도 동시에 재할당 되는것 같구요. 추가로 현재 controller에서 관리되고 있는 상태들도 초기화되지 않나 생각됩니다!

sooyeoniya replied:
헉.. 헤더 푸터와 같은 불필요한 렌더링이 있군요 🥲 생각이 좀 짧았던 부분이었네요. 다시 고려해보겠습니다!

sooyeoniya replied:
확실히 header나 footer는 정적 요소이기 때문에 굳이 리렌더링이 필요없는 것 같아서, main 태그 내부를 따로 main-container라는 명칭으로 커스텀 컴포넌트화 해서 해당 부분만 다시 렌더링되도록 해주었습니다. 이 과정에서 main-container에서만 관리되는 화면 렌더링 관련 메서드들도 main-container 내부로 옮겨서 캡슐화해줬습니다~!
https://github.com/woowacourse/javascript-lotto/pull/406/commits/ff1b8d035960f732143c4f152e0abe1fe81326d2

JUDONGHYEOK reviewed (changes requested):
안녕하세요! 니야! 마지막 연휴에 니야의 리뷰로 연휴를 마무리하게 되어서 영광입니다!0️⃣ 

 전체적으로 커스텀엘리먼트와 이벤트를 적절히 활용했다고 생각해서 재미있게 리뷰할 수 있었어요! 특히 BaseWebComponent를 만들어 공통 로직들을 재사용 가능하게 설계한 건 코드 일관성과 유지보수성을 높였다고 생각하구요, 커스텀 엘리먼트를 활용해 모듈화와 캡슐화를 추구한 점도 칭찬하고 싶습니다! 폴더 구조에서 컴포넌트를 독립적으로 분리한 것도 코드 볼륨 관리와 책임 분리에 큰 도움이 된 것 같아요.

이와 관련되어 개선할점이라고 생각될 부분들은 코멘트로 남겨두었어요! 다만 지금 첨부해주신 데모페이지가 정상적으로 동작하고 있지 않은 것 같은데 이것 한번 확인해주시면 감사하겠습니다!!

### 논의하고 싶은 부분
## 1️⃣ 컴포넌트 명, 이벤트 타입 명 피드백
딱히 어색한 부분은 보이지 않았던 것 같아요! 굳이굳이 뽑자면 result정도인데 사실 문제 없다고 생각합니다
## 2️⃣ UX 측면 피드백
UX도 좋았던 것 같은데요. 특히 에러메세지를 단순  alert로 표시하기보다 input창 근처에 표시하고 당첨번호와 같은경우에도 두자릿수 이상이면 다음 Input창으로 넘어가는 부분들이 인상적이였던 것 같아요! 
## 3️⃣ View 객체의 위치

views/web/View.js를 독립적으로 둔 건 MVC 패턴을 유지하려는 의도가 잘 반영된 선택이라고 생각합니다. components 폴더에 넣는 건 코드 볼륨 관리 측면에서 설득력 있다고 생각하지만, View가 전체 UI 구조를 책임진다는 점에서 독립적인 위치가 더 적절해 보여요. BaseWebComponent 상속 여부는 render만 사용하는 경우 굳이 필요 없으니, 현재처럼 별도 클래스로 두는 게 맞는 판단이라고 생각해요! View에 추가적인 역할을 부여할 수도 있을 것 같은데요. 이는 코멘트로 남겨두었습니다!

## 4️⃣ 전반적인 설계 방식과 흐름 피드백
MVC패턴 자체만 보면 controller의 역할에서 상태를 관리하는 것이 어색하긴하지만 전반적으로 customElement를 잘활용하신 것 같다는 생각입니다! customElement를 처음 사용하신게 맞나 싶을 정도였어요! 피드백을 드리자면 개인적으로 렌더링 위치는 View에서 중앙 관리하면 성능과 흐름이 더 명확해질 거라고 생각이 들구요.
 아무래도 disconnectedCallback가 restart한정해서 호출되기 때문에 컴포넌트가 DOM에 언제 포함되어야되고 안 포함되어야할지에 대한 로직이 추가적으로 포함되면 더 좋지 않을까 라는 생각인데 이건 선택적으로 반영해주셔도 됩니다! 또 성능 관련해서 shadowDOM을 시도해볼 수도 있을 것 같은데 이것 역시도 참고만 해주세요!

## 5️⃣ 상수화 고민
태그명 상수화는 LottoController에서 반복 접근 시 유용하니 적절한 선택이라고 생각해요. 하지만 니야의 생각처럼 클래스 선택자(lotto-purchase__form 등)까지 상수화하는 건 과도하다고 생각합니다.
 실무에서는 보통 재사용 빈도와 가독성 균형을 맞춰 상수화합니다. 예를 들어, LottoController처럼 여러 곳에서 참조되는 값(태그명, 이벤트 타입)은 상수화하고, 컴포넌트 내부에서만 쓰이는 클래스명은 그대로 둬도 괜찮아요. 가독성이 떨어질 우려가 크다면 현재처럼 최소화하는 게 맞습니다.

우선 데모페이지가 정상적으로 동작하는 것 같지 않아 RequestChanges하겠습니다! 현재 구조도 충분히 좋은 것 같아 남긴 리뷰는 필수가 아니라 참고차 받아드려주세요! 수고많으셨어요 니아!!

sooyeoniya commented:
동키콩! 데모 페이지 관련 문제 바로 해결했습니다 ! 확인 부탁드려용 :)

JUDONGHYEOK commented:
아 추가로 customElement를 처음 적용해보셨다고 하셨는데 적용하시기 전의 의도가 구현이 되었는지 customElement를 사용했을 때 니야가 느낀 장단점이 궁금해요!

sooyeoniya committed:
fix: favicon 배포 시 적용 안됨 이슈 수정

sooyeoniya committed:
refactor: 웹 LottoController 상태 관리 외부 Domain으로 분리

- MVC 패턴에 따라 명확히 역할 구분을 하기 위해 controller에서 domain과 view를 외부에서 주입시킴
- domain은 기존에 controller에서 가지고 있던 lottos(발행된 로또 배열들)와 purchaseAmount(로또 구입 금액)에 대한 상태 관리 책임을 대신 맡음
- LottoDomain은 WinningStatistic 라는 도메인 객체를 조합(Composition)하여 웹을 위한 도메인 객체를 구성함
- View 관련 메서드는 모두 Controller -> View 로 이동함 (DOM 조작 캡슐화)
- 이로써, Controller를 좀 더 순수하게 이벤트 처리와 domain과 view의 중재자 역할만 수행하도록 함

sooyeoniya committed:
refactor(domains): issueLottos, WinningStatistics 파일 common 폴더로 이동

- web 폴더와 같은 depth에 두기 위함
- 파일 역할에 맞춰 폴더를 명확히 분리하기 위함

sooyeoniya committed:
refactor(domUtils): querySelector, querySelectorAll 랩핑하는 유틸함수 제거

- 유틸화에 대하여 오히려 정보를 은닉하여 해당 유틸이 querySelector라는 것을 제대로 알 수 없게 함
- 참조를 하나 더 생기는 느낌의 유틸함수여서 큰 이점이 없다고 판단함

sooyeoniya committed:
refactor: WinningLotto와 IssuedLotto 컴포넌트의 플래그 변수 제거 및 View에서 컴포넌트 렌더링 여부 책임 위임

- WinningLotto와 IssuedLotto 컴포넌트 내부에 있는 플래그 변수 제거
- View에서 domUtils에 있는 hideElement와 renderElement를 사용해 본 컴포넌트의 렌더링 역할을 담당하도록 함
- 처음 view가 그려질 때 각 컴포넌트에 hidden 클래스를 달아 화면에서 숨김
- controller에서 view의 updateIssuedLotto(발행된 로또 정보 업데이트), initWinningLotto(당첨 번호 및 보너스 번호 렌더링)가 호출될 때 hidden 클래스를 제거하여 화면에 두 개의 컴포넌트들이 나타나도록
- 컴포넌트의 렌더링 여부를 컴포넌트 내부에서 조작하면 렌더링 정보가 은닉되어 로직 파악이 어려움
- 따라서 controller와 view에서 관리하도록 수정함

sooyeoniya committed:
refactor(View): 로또 재시작 시 불필요한 렌더링 요소 줄이기 위해 컴포넌트 관심사 분리

- main container 내부를 header, footer와 분리하여 로또 재시작 시 main container 내부만 리렌더링 되도록 수정함
- MainContainer 커스텀 컴포넌트를 정의하여 관리함.
- 또한 issuedLotto와 winningLotto 컴포넌트 렌더링 제어 부분도 main container로 옮겨 스스로 직접 렌더링 관리하도록 수정

sooyeoniya committed:
refactor(View): querySelector로 선택한 요소 캐싱하여 재사용하도록 수정

- mainContainer 요소가 querySelector로 2번 호출되는 현상 있었음
- 호출 최소화를 위해 요소 캐싱하여 재사용하도록 함

sooyeoniya committed:
refactor(styles): 본 프로젝트에 더 효율적인 스타일 로드 방식 적용

- CSS 내부의 @import 방식은 link 방식보다 로드 시간 긺 (link 방식은 여러 CSS 파일을 병렬 처리)
- JS 파일 내부에서 global.css 를 import 로 호출하는 것보다 link 태그를 사용하는 것이 FOUC 문제 방지 가능
- link 태그는 HTML에서 CSS 함께 로드하기 때문에 스타일 빠르게 적용 가능
- link 태그방식은 CSS 가 모든 페이지에 로드되기 때문에 전역 스타일 적용에 적합

결론: global.css, reset.css 를 모두 link 태그 방식으로 변경함

sooyeoniya reviewed (commented):
안녕하세요, 동키콩! 니야입니다 :)
이번주 일정 때문에 리뷰 반영이 조금 늦어졌네요 🥲
기다리게 해서 죄송합니다 !! 동키콩이 리뷰해주신대로 좀 더 깊이 있게 고민해보고 열심히 반영해봤어용 😁
아래는 리뷰 반영 요약표입니다 ! 확인 부탁드려요 👍🏻 

## ✅ 리뷰 반영 요약표
내용 | 관련 코멘트 | 커밋
-- | -- | --
배포 시 favicon 적용 안됨 이슈 | [favicon 관련 코멘트](https://github.com/woowacourse/javascript-lotto/pull/406#discussion_r1977491763) |  https://github.com/woowacourse/javascript-lotto/pull/406/commits/892ac102dd0aab6e611c5beb5f1010a7b236ccf0
도메인 관련 상태관리 및 MVC 역할 재정의 | [MVC 각 역할 문제 및 상태 관리 관련 코멘트](https://github.com/woowacourse/javascript-lotto/pull/406#discussion_r1977320876)  |  https://github.com/woowacourse/javascript-lotto/pull/406/commits/7ef9aeec3098e3f6a81f862dbfb9d2b1114a8bfb
 querySelector 랩핑한 유틸함수 제거 | [querySelector 유틸함수 관련 코멘트](https://github.com/woowacourse/javascript-lotto/pull/406#discussion_r1977440860) |  https://github.com/woowacourse/javascript-lotto/pull/406/commits/3e6a90857af10d3d1af3a79f53c6ab706baef90d
플래그 변수 제거 및 렌더링 역할 외부로 분리 | [플래그 변수 관련 코멘트](https://github.com/woowacourse/javascript-lotto/pull/406#discussion_r1977470416) |  https://github.com/woowacourse/javascript-lotto/pull/406/commits/a22f4547ccf4cf6db8671b59f600355e5a99f06f
정적 요소의 불필요한 렌더링 이슈 | [로또 재시작 비용 관련 코멘트](https://github.com/woowacourse/javascript-lotto/pull/406#discussion_r1977542521) |  https://github.com/woowacourse/javascript-lotto/pull/406/commits/ff1b8d035960f732143c4f152e0abe1fe81326d2
더 효율적인 CSS 파일 로드 방식 | [CSS 파일 로드 방식 차이](https://github.com/woowacourse/javascript-lotto/pull/406#discussion_r1977314693) | https://github.com/woowacourse/javascript-lotto/pull/406/commits/ffdb580572e8901277191735a5f11b547f3846c3

## 🐢 `customElement`에 대한 답변

우선 UI 컴포넌트를 잘 분리해서 유지보수를 쉽게 하려면 어떻게 하면 좋을지 계속 설계 관점에서 찾아봤던 것 같습니다. 그 과정에서 customElement를 발견했고 그걸 도입해보고자 하였습니다! 

좋았던 점은 확실히 컴포넌트 캡슐화가 뛰어나고 관리가 편하다는 장점이 있습니다! 컴포넌트 내부에서 필요한 메서드들을 추상 클래스에서 공통으로 정의해놓고 사용하다보니, 훨씬 유지보수하기도 좋고 코드의 가독성 측면에서도 보기 좋았습니다.

다만, customElement를 사용하니까, 부모에서 이벤트를 잡고 그때 사용할 정보들을 함께 넘겨주기 위해 CustomEvent를 사용해서 수동으로 직접 이벤트 버블링을 시켜줘야 제대로 동작하더라구요. 그리고 customElement를 사용하면서 shadowDOM 을 함께 적용해보려고 했는데 전역 CSS를 상속받지 못해서, 각 컴포넌트 내부에서 별도로 전역 스타일을 적용해줘야하는 번거로움이 있었습니다. 저는 전역 스타일들(global.css, reset.css)을 각 컴포넌트에 적용해줘야했기 때문에 shadowDOM을 따로 사용하지 않았습니다. 추후에 기회가 된다면 shadowDOM 을 사용해서 더 캡슐화된 방식으로 한 번 구현해보겠습니다!

덕분에 customElement에 대해서 다시 한 번 곱씹어 볼 수 있었습니다! 좋은 질문 감사합니다 :)

JUDONGHYEOK commented on src/components/header/header.css:
```diff
@@ -0,0 +1,14 @@
+  align-items: center;
+  padding-left: 130px;
+  background-color: var(--lotto-primary-color);
+  position: fixed;
```
아 제가 놓친 부분인데 이녀석 때문에 모바일에서 화면이 이상하게 나오고 있었네요. 참고만 해주세요!

JUDONGHYEOK reviewed (approved):
안녕하세요! 니야~~ 2단계 미션도 너무 고생많으셨어요.
리뷰들 하나하나 꼼꼼히 답변해주셔서 너무 감동이였어요🥹
니야 덕분에 저도 많이 배워갔던 미션이였던 것 같아요.

앞으로 미션을 수행하시면서 문제에 맞는 해결책을 계속 의식하시면서 연습하시면 더 좋을 것 같다는 생각이 들어요!

너무너무 수고많으셨고 이어지는 다음 미션들도 화이팅입니다!!
